### PR TITLE
Lesson 19

### PR DESCRIPTION
**Pull request by vtx123**

### CODE REVIEW

File: Lesson19/DynamicIntArray.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// Copy c-tor
+DynamicIntArray::DynamicIntArray(const DynamicIntArray& other) {
+    if (m_size < other.getSize()) {
+        delete[] m_elements;
+        m_size = other.m_size;
+    }
+
+    m_elements = new int[m_size];
```
Comment (helen-black) [inline Lesson19/DynamicIntArray.cpp]: memory leak

File: Lesson19/DynamicIntArray.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    delete[] m_elements;
+}
+
+DynamicIntArray& DynamicIntArray::operator=(const DynamicIntArray& other) {
+    return *this;
+}
+
+int& DynamicIntArray::operator[](std::size_t index) {
+    return m_elements[index];
+}
```
Comment (helen-black) [inline Lesson19/DynamicIntArray.cpp]: validation is needed at least for DEBUG version

File: Lesson19/DynamicIntArray.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
+#include "DynamicIntArray.h"
+#include <iostream>
+
+// Default c-tor
+DynamicIntArray::DynamicIntArray() :m_size(5), m_position(0) {
+    m_elements = new int[m_size];
+}
```
Comment (helen-black) [inline Lesson19/DynamicIntArray.cpp]: if you're doing in this way, you should rename m_size to m_capacity, m_position to m_size

File: Lesson19/DynamicIntArray.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+DynamicIntArray::DynamicIntArray(const DynamicIntArray& other) {
+    if (m_size < other.getSize()) {
+        delete[] m_elements;
+        m_size = other.m_size;
+    }
+
+    m_elements = new int[m_size];
+    for (int i = 0; i < m_size; i++) {
+        m_elements[i] = other.m_elements[i];
+    }
```
Comment (helen-black) [inline Lesson19/DynamicIntArray.cpp]: you can use std::copy here

File: Lesson19/DynamicIntArray.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// Free allocated memory
+DynamicIntArray::~DynamicIntArray() {
+    delete[] m_elements;
+}
+
+DynamicIntArray& DynamicIntArray::operator=(const DynamicIntArray& other) {
+    return *this;
+}
```
Comment (helen-black) [inline Lesson19/DynamicIntArray.cpp]: it doesn't work well.
you should remove your data, and allocate memory for a new array to store other.m_elements

File: Lesson19/Lesson19.cpp [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    }
+    std::cout << std::endl << std::endl;
+
+    const int arr2NewSize = 15;
+    arr2.setSize(arr2NewSize);
+    for (int i = 0; i < arr2NewSize; i++) {
+        arr2.push_back(i + 10);
+        std::cout << "[" << arr2[i] << "] ";
+    }
+    std::cout << std::endl << std::endl;
```
Comment (helen-black) [inline Lesson19/Lesson19.cpp:31]: check output, please. poes it work well?

Comment (helen-black) [reply Lesson19/Lesson19.cpp:31]: at the end it should print
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14

**Review by helen-black — CHANGES_REQUESTED** [review]

File: Lesson19/DynamicIntArray.cpp (hunk 1 above)
Comment (vtx123) [reply Lesson19/DynamicIntArray.cpp]: fixed

File: Lesson19/DynamicIntArray.cpp (hunk 2 above)
Comment (vtx123) [reply Lesson19/DynamicIntArray.cpp]: added.

File: Lesson19/DynamicIntArray.cpp (hunk 3 above)
Comment (vtx123) [reply Lesson19/DynamicIntArray.cpp]: ok, i removed 5. now default value is 0.

File: Lesson19/DynamicIntArray.cpp (hunk 4 above)
Comment (vtx123) [reply Lesson19/DynamicIntArray.cpp]: done thx.

File: Lesson19/DynamicIntArray.cpp (hunk 5 above)
Comment (vtx123) [reply Lesson19/DynamicIntArray.cpp]: redone.

File: Lesson19/Lesson19.cpp (hunk 6 above)
Comment (vtx123) [reply Lesson19/Lesson19.cpp:31]: fixed, you right i misunderstood description of HW.

File: Lesson19/DynamicIntArray.cpp [hunk 7]
```diff
@@ -2,53 +2,63 @@
...
-    return *this;
+    DynamicIntArray tmp(other);
+    delete[] m_elements;
+    return tmp;
 }
 
 int& DynamicIntArray::operator[](std::size_t index) {
+    if (index >= m_size) {
+        index = m_size - 1;
+    }
```
Comment (helen-black) [inline Lesson19/DynamicIntArray.cpp]: it's wrong behaviour. if index is out of range you should throw an exception or use assert

File: Lesson19/DynamicIntArray.cpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            tmp[i] = m_elements[i];
+        }
+        tmp[m_position] = element;
+        m_size = m_position + 1;
+        delete[] m_elements;
+        m_elements = new int[m_size];
+        for (int i = 0; i < m_size; i++) {
+            m_elements[i] = tmp[i];
+        }
+        delete[] tmp;
```
Comment (helen-black) [inline Lesson19/DynamicIntArray.cpp]: m_elements = tmp;

File: Lesson19/DynamicIntArray.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+// Allocate new memory for new element
+// Copy old content to new inner array
+// insert element at the last index
+// update size
+void DynamicIntArray::push_back(int element) {
+    if (m_position + 1 >= m_size) {
+        int* tmp = new int[m_position + 1];
+        for (int i = 0; i < m_size; i++) {
+            tmp[i] = m_elements[i];
+        }
```
Comment (helen-black) [inline Lesson19/DynamicIntArray.cpp]: std::copy

File: Lesson19/DynamicIntArray.cpp [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        m_elements = new int[m_size];
+        for (int i = 0; i < m_size; i++) {
+            m_elements[i] = tmp[i];
+        }
+        delete[] tmp;
+    }
+    else {
+        m_elements[m_position] = element;
+    }
+    m_position++;
```
Comment (helen-black) [inline Lesson19/DynamicIntArray.cpp:98]: if you analyse code of your class, you can understand that you don't really need this m_position variable

**Review by helen-black — CHANGES_REQUESTED** [review]

File: Lesson19/DynamicIntArray.cpp (hunk 7 above)
Comment (vtx123) [reply Lesson19/DynamicIntArray.cpp]: added assert

File: Lesson19/DynamicIntArray.cpp (hunk 8 above)
Comment (vtx123) [reply Lesson19/DynamicIntArray.cpp]: fixed

File: Lesson19/DynamicIntArray.cpp (hunk 9 above)
Comment (vtx123) [reply Lesson19/DynamicIntArray.cpp]: fixed

File: Lesson19/DynamicIntArray.cpp (hunk 10 above)
Comment (vtx123) [reply Lesson19/DynamicIntArray.cpp:98]: yees;( I will fix it in lesson 22

**Pull request merged**